### PR TITLE
AST_to_IL: python: Translate calls to constructors as `new` instructions

### PR DESCRIPTION
**Pull request by akuhlens**
This PR extends the AST_to_IL transformation to handle calls that have been resolved to python constructors to class creation via New. This adds support for field sensitivity in Python class constructors.
 
This PR also adds a List_ utility function to get the last element of a list.

[Corresponding PR in Semgrep Pro](https://github.com/semgrep/semgrep-proprietary/pull/1287)

# Test Plan
See semgrep pro PR.

### CODE REVIEW

File: libs/commons/List_.ml
Comment (nmote) [inline libs/commons/List_.ml]: I'd rather not encourage people to use exceptions. Just implement last_opt directly and leave last_exn out.

**Review by nmote — COMMENTED** [review]
I'm not super familiar with AST_to_IL, but this looks to me like it will only handle statements of the form `x = Y()`. Something like `foo(Y())` would still have the old, incorrect behavior. Do I understand this correctly? How difficult would it be to generalize this to any constructor call within an expression, no matter the position?

**Comment by akuhlens** [comment]
> I'm not super familiar with AST_to_IL, but this looks to me like it will only handle statements of the form `x = Y()`. Something like `foo(Y())` would still have the old, incorrect behavior. Do I understand this correctly? How difficult would it be to generalize this to any constructor call within an expression, no matter the position?

I do think your understanding of the code is correct, and it isn't impossible to fix, though I do think we would need to patch up the type information for `__init__` so that we have access to the return type which we are currently getting from the lhs variable. I wonder if path sensitivity works for `foo((new Y()).field)` I don't think it does but I will double check.

(edit / update)

Indeed I added a new test to the pro repo that shows that `foo((new Y()).field)` for java doesn't currently work. I would prefer to land this and then worry about the general fix to field sensitivity which should be possible.

File: src/analyzing/AST_to_IL.ml
Comment (nmote) [inline src/analyzing/AST_to_IL.ml]: I'd rather pattern match on `lhs_info` and have the pattern match duplicated in both cases where that is bound than have a failwith here.

**Review by nmote — APPROVED** [review]
This looks good, but I'm not an expert here so I'd prefer to give @IagoAbal a chance to take a quick look on Monday. Let's hold off on merging until then.

File: src/analyzing/AST_to_IL.ml
Comment (IagoAbal) [reply src/analyzing/AST_to_IL.ml]: Definitely no `failwith` allowed...

**Review by IagoAbal — CHANGES_REQUESTED** [review]
Just making sure I review this before it gets merged.

File: libs/commons/List_.ml
Comment (IagoAbal) [inline libs/commons/List_.ml:222]: Despite you checked the box I see no changelog.

**Comment by IagoAbal** [comment]
> Indeed I added a new test to the pro repo that shows that `foo((new Y()).field)` for java doesn't currently work. I would prefer to land this and then worry about the general fix to field sensitivity which should be possible.

Don't worry about that case now, I think I'll make some progress towards supporting that and other stuff in the following weeks.

File: src/analyzing/AST_to_IL.ml
Comment (IagoAbal) [inline src/analyzing/AST_to_IL.ml]: This comment seems to have been copied, in the original place it makes sense and it should remain, but here we would need another comment. Just explain that we encode `x = new T(args)` as `x = new x.T(args)` so that taint analysis knows that the `this` when calling `T` is the variable `x`. It's kinda tacky but works for now.

File: src/analyzing/AST_to_IL.ml
Comment (IagoAbal) [inline src/analyzing/AST_to_IL.ml]: Pass the name `obj'` to `mk_class_construction` rather than `lval`, that code is assuming that we have a `x = new T()` kind of statement, it's not correct if the LHS is not a simple variable so we shouldn't be passing it a (potentially) arbitrary l-value..

File: src/analyzing/AST_to_IL.ml
Comment (aryx) [inline src/analyzing/AST_to_IL.ml]: a match when match ... is a bit heavy. You probably want to move this code in an helper function like
is_python_init_init (),  or even is_constructor_call()

File: src/analyzing/AST_to_IL.ml
Comment (IagoAbal) [inline src/analyzing/AST_to_IL.ml]: We don't support this case yet in your `mk_class_construction` helper, for now just `x = new ...`, where `x` is a simple variable. I think it's better to have a separate PR to allow `DotAccess` and then we want to do it not only here but also for Java & co.

File: src/analyzing/AST_to_IL.ml
Comment (IagoAbal) [inline src/analyzing/AST_to_IL.ml:711]: Well, also because the goal of the IL is also to make things "uniform" across languages. Add a `see HACK(new)` here so when reading this comment one can jump to the same code that we use for languages with explicit `new`.

File: src/analyzing/AST_to_IL.ml
Comment (IagoAbal) [reply src/analyzing/AST_to_IL.ml]: Totally agree with Pad here!

File: libs/commons/List_.ml
Comment (akuhlens) [reply libs/commons/List_.ml:222]: This diff alone doesn't change any user facing behavior, so I wasn't sure that it required, but I will add one.

File: src/analyzing/AST_to_IL.ml
Comment (akuhlens) [reply src/analyzing/AST_to_IL.ml]: I am not sure what would need to change to support it, but I will remove it. I thought I abstracted over the various callees by calling `lval` on `lhs_e` but I will trust your judgement and remove this case.

File: changelog.d/ea-272.added
Comment (IagoAbal) [inline changelog.d/ea-272.added]: I'd typically add `Pro:` in front to clarify this is about Pro.

File: src/analyzing/AST_to_IL.ml
Comment (IagoAbal) [inline src/analyzing/AST_to_IL.ml]: This comment is very confusing, why is it "too loose", how could we "tighten this up" ???

File: src/analyzing/AST_to_IL.ml
Comment (IagoAbal) [inline src/analyzing/AST_to_IL.ml:278]: Avoid `match` nested in a `when`, better define helper.

Comment (IagoAbal) [reply src/analyzing/AST_to_IL.ml:278]: My bad, no `when` in here.

File: src/analyzing/AST_to_IL.ml
Comment (IagoAbal) [inline src/analyzing/AST_to_IL.ml:287]: Add a comment to clarify what this means, I'm guessing, the result type is a "type name", presumably the name of a class?

File: src/analyzing/AST_to_IL.ml
Comment (IagoAbal) [inline src/analyzing/AST_to_IL.ml:731]: Can this be a `DotAccess` ? In which case ? A qualified name ? Good to clarify in a comment.

File: src/analyzing/AST_to_IL.ml
Comment (IagoAbal) [inline src/analyzing/AST_to_IL.ml]: Typo: "hacky"

**Review by IagoAbal — APPROVED** [review]
Nice!

**Pull request merged**
